### PR TITLE
chore(FDS-442): [prop-types] - Add CJS and ESM builds

### PR DESCRIPTION
**Pull request by lordpixel**
Resolves [FDS-442](https://espressive.atlassian.net/browse/FDS-442).

This updates the `rollup.config.js` file at the root of the monorepo, it allows for more controlled build process for packages other than `@espressive/cascara` that might not have scss or private files such as `@espressive/prop-types`.

### Dependencies
Changes to package.json file to add scripts and `walk-sync` dependency.

## PR Author Checklist

- [x] PR title adheres to [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#specification)
- [x] I have finished reviewing the contents of this PR for accuracy
- [x] If needed, I have included a [Changeset](https://github.com/changesets/changesets) and it follows [Semantic Versioning principles](https://semver.org/)
- [x] I have completed all of the above and have enabled `auto-merge` for this PR

### CODE REVIEW

File: utils/prop-types/package.json [hunk 1]
```diff
@@ -9,16 +9,25 @@
   },
   "license": "MIT",
   "author": "Rocio Baena <rocio.baena@espressive.com>",
+  "main": "dist/index.js",
   "module": "src/index.js",
```
Comment (brianespinosa) [inline utils/prop-types/package.json:13]: We should fix the `main` and `module` paths to be the same as they are in the Cascara package as these values are used directly in the rollup config to know where to output files.

If the Rollup config was run with these values in the package.json, you may have seen the files in src getting overwritten. I would even take a look at the index.js file there specifically as the Rollup config should be looking at this and it is essentially going to try to output esm directly into the src directory with this.

File: utils/babel-preset-espressive/package.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
     "src"
   ],
   "scripts": {
-    "prepare": "echo 'No prepare scripts for this package'"
+    "prepare": "echo 'No prepare scripts for babel-preset-espressive'"
```
Comment (brianespinosa) [inline utils/babel-preset-espressive/package.json:16]: I like this. Thanks for making these strings more specific!

File: rollup.config.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ import { nodeResolve } from '@rollup/plugin-node-resolve';
 import postcss from 'rollup-plugin-postcss';
 import stringHash from 'string-hash';
 import cssUrl from 'postcss-url';
+import walkSync from 'walk-sync';
```
Comment (brianespinosa) [inline rollup.config.js:8]: I have a few concerns about us adding an extra dependency here. We should in theory just be able to use `fs.readdirsync` from node to do all of this. In fact... there are kind of examples of how this would work in our Docs, as we are using just basic node functions to crawl our repo directories and find the MDX files. Is there a strong reason we are adding another dependency here instead of just using node?

File: rollup.config.js [hunk 4]
```diff
@@ -67,13 +68,38 @@ const external = (id) =>
...
+    walkSync(SOURCE_DIR, {
+      globs: ['private.js'],
+    }).length > 0;
+
   // Get the package.json file
   const pkgConfig = require(`${SOURCE_DIR}/package.json`);
+
   // Relative input location for Rollup to bundle from
-  const input = [`${SOURCE_DIR}/src/index.js`, `${SOURCE_DIR}/src/private.js`];
+  const input = [`${SOURCE_DIR}/src/index.js`];
```
Comment (brianespinosa) [inline rollup.config.js:86]: It looks like there are some pretty big changes in this file... but I am pretty sure the cause of our rollup config not working on everything else is this line here where we have the private entry.

I am pretty sure Rollup will just not use any of the other configured plugins if it does not encounter any files of those types. I suspect we might just be able to update this one thing and have a conditional input for the private exports in Cascara.

Were we seeing errors about the CSS plugins not being used? We could likely just check for private exports and call it a day.

File: utils/prop-types/package.json (hunk 1 above)
Comment (lordpixel) [reply utils/prop-types/package.json:13]: working on it 👍

File: rollup.config.js (hunk 3 above)
Comment (lordpixel) [reply rollup.config.js:8]: makes sense, working on it.

File: rollup.config.js (hunk 4 above)
Comment (lordpixel) [reply rollup.config.js:86]: We saw some errors regarding rollup css plugins in `@espressive/prop-types` package. That's why we changed it not to use those that plugin if it's not needed.

File: rollup.config.js (hunk 3 above)
Comment (lordpixel) [reply rollup.config.js]: Dependency removed.

File: utils/prop-types/package.json (hunk 1 above)
Comment (lordpixel) [reply utils/prop-types/package.json]: This was updated as requested

File: rollup.config.js (hunk 4 above)
Comment (lordpixel) [reply rollup.config.js:77]: ok, I tried reverting back to what it was and I didn't see the errors anymore, we are keeping all plugins.

**Comment by lordpixel** [comment]
@brianespinosa I think the build should work now, but there are some issues with the cache (I think).

File: rollup.config.js [hunk 5]
```diff
@@ -65,16 +66,32 @@ const external = (id) =>
...
+    input.push(PRIVATE_PATH);
+  }
 
-  // Shared Rollup plugins
+  // Shared Rollup plugins for code-only packages
   const rollupPlugins = [nodeResolve(), postcss(getPostCSSOptions()), json()];
 
+  // Gather info about the package
+  const IS_CJS = Boolean(pkgConfig?.main);
+  const IS_ESM = Boolean(pkgConfig?.module);
```
Comment (brianespinosa) [inline rollup.config.js:89]:
```suggestion
  const IS_ESM = pkgConfig?.type === 'module' || Boolean(pkgConfig?.module);
```
Technically we should also be checking the type first to determine if it is ESM. If there is a type then, the package should definitely be a module. Additionally, if `type === 'module'` then we should not have any CJS at all. So that would be the best way to check if ESM_ONLY. Any package that is only a module using type should not have CJS.

**Review by rousspb — APPROVED** [review]

**Review by rociobaena — DISMISSED** [review]

File: .changeset/wild-kids-buy.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+"@espressive/prop-types": patch
+---
+
+chore(FDS-442): [Rollup] - Improve rollup config to handle code-only packages
```
Comment (brianespinosa) [inline .changeset/wild-kids-buy.md:5]:
```suggestion
chore(FDS-442): [@espressive/prop-types] - Added a main (cjs) and modules (esm) entry point to the package
```
If this change was only to the rollup config, we would not need a Changeset because we do not ship the Rollup config. However, this update to the config changes the prop-types package and the changeset should describe the impact of that change.

File: rollup.config.js [hunk 7]
```diff
@@ -65,16 +66,32 @@ const external = (id) =>
...
-  const input = [`${SOURCE_DIR}/src/index.js`, `${SOURCE_DIR}/src/private.js`];
+  const input = [`${SOURCE_DIR}/src/index.js`];
+
+  // Not all packages have private exports
+  if (fs.existsSync(PRIVATE_PATH)) {
+    input.push(PRIVATE_PATH);
+  }
 
-  // Shared Rollup plugins
+  // Shared Rollup plugins for code-only packages
```
Comment (brianespinosa) [inline rollup.config.js:84]:
```suggestion
  // Shared Rollup plugins
```
This should go back to the previous comment. This change only has to do with detecting if there is a private input/output.

File: rollup.config.js [hunk 8]
```diff
@@ -65,16 +66,32 @@ const external = (id) =>
...
   const rollupPlugins = [nodeResolve(), postcss(getPostCSSOptions()), json()];
 
+  // Gather info about the package
+  const IS_CJS = Boolean(pkgConfig?.main);
+  const IS_ESM = pkgConfig?.type === 'module' || Boolean(pkgConfig?.module);
+  const TYPE = pkgConfig?.type;
+
+  // check if only ESM is part of the output
+  const ESM_ONLY = IS_CJS && !IS_ESM && TYPE === 'module';
```
Comment (brianespinosa) [inline rollup.config.js:93]: This logic is still inconsistent. Will revise to align with the presence of "module" or not.

**Review by brianespinosa — CHANGES_REQUESTED** [review]
Updating this PR with items to change so I can go in an get these last few things cleaned up and bring this in since Manu is out today. 😷

File: rollup.config.js [hunk 9]
```diff
@@ -100,7 +117,8 @@ const getRollupConfig = ({ pwd, babelConfigFile }) => {
     external,
     input,
     output: {
-      dir: `${SOURCE_DIR}/${pkgConfig.main.replace('/index.js', '')}`,
```
Comment (brianespinosa) [inline rollup.config.js:103]: We do not want to hard code the dir here because someone might want to configure a different directory for these files.

File: rollup.config.js [hunk 10]
```diff
@@ -67,10 +68,18 @@ const external = (id) =>
...
   const pkgConfig = require(`${SOURCE_DIR}/package.json`);
+
   // Relative input location for Rollup to bundle from
-  const input = [`${SOURCE_DIR}/src/index.js`, `${SOURCE_DIR}/src/private.js`];
+  const input = [`${SOURCE_DIR}/src/index.js`];
+
+  // If we have a private export, add it to our list of inputs
+  if (fs.existsSync(PRIVATE_PATH)) {
+    input.push(PRIVATE_PATH);
+  }
```
Comment (brianespinosa) [inline rollup.config.js:82]: After tests rolling back a few other things, this is really the only change that was needed. This rollup config should work with anything that has proper entry points defined now. The only edge case was us having a private export, which is now supported with this check. Everything else works correctly as it was (minus a discovery of an error about exports: auto in CJS)

**Review by brianespinosa — APPROVED** [review]
